### ISSUE TITLE
Sometimes, delete() hangs MATLAB

### ISSUE THREAD

**Issue by cntaylor** [DylanMuir/ParforProgMon#4]
Great code, FYI...

I have some code that opens and closes a window.  In general, something like this:

```
gcp; %Make sure the parallel pool is started
num_mc = 1000;
for iter=1:3
    window_title = sprintf('MC runs, iter %d: ',iter);
    ppm = ParforProgMon(window_title, num_mc);
    parfor mc = 1:num_mc
        %Some Monte Carlo (MC) code here...
        ppm.increment();
    end
    %some outer iteration code here...
    ppm.delete();
end
```

About one out of 10 times, ppm.delete() will hang MATLAB.  I cannot Ctrl-C out, or pause in debugging mode.  The only thing that works is to kill MATLAB and restart it.  I put a printf statement before and after ppm.delete() and when it crashed now, I get the before but not after, so I am pretty sure it is in the ppm.delete() code that it is hanging.  Not really any more details than this though.  Sorry!

OS:   Windows 10
MATLAB: 2017b

Thanks!

**Comment by DylanMuir** [comment]
Hi, thanks for your bug report! Unfortunately I don’t have access to a Windows 10 machine with Matlab at the moment, so this will be difficult for me to debug.

You could put some print statements in `ParforProgMon.delete()` or in the `done()` method in the Java class (as long as you can recompile). That might give you some more hints as to what’s going on.

**Comment by cntaylor** [comment]
Looking at the code (as you suggested), it gets into the if statement in delete( o ) in the MATLAB code.  Something about `o.JavaBit.done()` kills it.  Looking at the code though, realized that the `.done()` is already being called in the Java code when `fStep*newVal >= fBar.getMaximum()`.  So, I didn't even need the .delete() call.  Once I remove that, it works.

I am guessing it doesn't like a call to .done() if a previous call has already been made.  Not familiar enough with Java in MATLAB though to know how to fix it.

**Comment by DylanMuir** [comment]
Thanks for looking into this, that’s really helpful! I can try hacking together a fix, and I’ll let you know when I’ve got something that should work.

**Comment by cntaylor** [comment]
I had a thought this morning and tried it out.  If I just do a pause(.01) in MATLAB before calling `delete()`, then I don't have this problem any more.  I would suggest the following modification to delete() in ParforProgMon.m
```
      function delete( o )
         % delete - METHOD Delete a ParforProgMon object

         %Make sure that any other threads that may have closed 
         %the UI down have a chance to do it first
         pause(.01)
         % Close the UI
         if (~isempty(o.JavaBit))
            o.JavaBit.done();
         end
      end
```

On my machine, this fixes the problem.

**Comment by DylanMuir** [comment]
Thanks very much for the bug fix! I've incorporated it into new branch: https://github.com/DylanMuir/ParforProgMon/tree/iss4
Please check out that branch and give it a go. Let me know if this works or fails for you, and I'll merge it.

**Comment by cntaylor** [comment]
It fixed it on my machine.  Thanks!

**Comment by DylanMuir** [comment]
Merged in #5. Thanks @cntaylor for your help!
